### DISCUSSION TITLE
What would it take to re-enable EAP-PEAP plugins for Android?

### DISCUSSION THREAD

**Discussion by lowjoel** [strongswan/strongswan#1676] [Q&A]
Hey there, I saw that eap-peap was disabled for Android: https://github.com/strongswan/strongswan/commit/ba9255b3586755603a2e6a3486d4dd94827ce7bd but there wasn't too much explanation as to what is missing. The commit message says:

> We have to make this optional and more configurable.  It seems some
> commercial VPN providers use self-signed certificates for their AAA
> servers.

But it doesn't say what would be the remedy for it. I'm thinking of trying to get this to work in the coming weeks, what should I work on?

**Comment by tobiasbrunner** [comment]
The problem with EAP-PEAP and EAP-TTLS is, or at least was back then, that some large commercial VPN providers request it as their initial EAP method but do so with problematic configurations (no idea what the situation would be now, nearly five years later). It would most likely have to be a separate authentication method that the user had to select manually (with a different set of plugins loaded by the native part of the app).

But even if the EAP methods are optional, there are still some issues. One is the identity of the AAA server, which is currently not configurable. So the behavior is currently similar to the NetworkManager client, where `%any` is used as AAA server identity, i.e. any trusted AAA server certificate will be accepted. While not ideal (making the identity configurable would definitely be preferable, even though it's an advanced option that most users won't know how to configure), this works but it still has two problems.

One is that the CA could be different from that used for the VPN server's certificate. So that would require the ability to select and load multiple CA/server certificates per connection, as it, otherwise, forces users to enable the automatic CA selection, which makes the app trust all locally available CA/server certificates, which some users might not want to do. (There is a similar issue in the NM client, see [here for a workaround](https://github.com/strongswan/strongswan/issues/1557#issuecomment-1446612861).)

The second problem is that some VPN providers used self-signed certificates for their AAA servers (possible that there are clients that accept arbitrary AAA server certificates). strongSwan will only accept those if they are locally installed and trusted, which is probably impossible for most users of commercial VPN services (this is the main reason why the _eap-peap_ and _eap-ttls_ plugins are currently disabled and will probably have to be made optional). And for advanced users that use such a setup, it would again require the ability to install and select an additional AAA server certificate.

The AAA identity and AAA server/CA certificate would also have to be configurable and importable via [VPN profile files](https://docs.strongswan.org/docs/5.9/os/androidVpnClientProfiles.html).

All in all it's quite some work for a feature that hasn't been requested much over the last years (there was the [ticket for EAP-TTLS](https://wiki.strongswan.org/issues/2392) in the old issue tracker in 2017, that triggered the changes you referred to - note that there was no feedback after reverting them -, and a [ticket for EAP-PEAP](https://wiki.strongswan.org/issues/3244) in 2019, nothing since then).
